### DISCUSSION TITLE
PEP-612 return parameter specification type hinting

### DISCUSSION THREAD

**Discussion by Mardoxx** [microsoft/pyright#5120] [General]
With Python 3.11

Taking the example here, https://peps.python.org/pep-0612/#semantics

```python
P = ParamSpec("P")
U = TypeVar("U")

class Y(Generic[U, P]):
  f: Callable[P, str]
  prop: U

  def __init__(self, f: Callable[P, str], prop: U) -> None:
    self.f = f
    self.prop = prop

def a(q: int) -> str: ...

Y(a, 1)   # Should resolve to Y[(q: int), int]
Y(a, 1).f # Should resolve to (q: int) -> str
```

What is this type definition here? `Y[(q: int), int]` - I understand what its intent is, but what _is_ it

`foo: Y[(q: int), int] = ...` gives a type error. So why does `foo = Y(a, 1)` display it instead of `Y[Unknown, int]` or something else

Being able to specify named arguments like this would be very useful, is it proposed?

**Comment by erictraut** [comment]
Converting to discussion since this is a question, not a bug report.

**Reply by erictraut** [reply]
Are you asking about the "revealed type" text (or "hover" text) string that pyright uses to represent a generic class that includes a ParamSpec that has been specialized?

A ParamSpec represents the input signature of a callable. It captures all aspects of the signature including the types of the parameters and their names. It even captures overloads (multiple signatures for the same function). 

Some ParmaSpec specializations cannot be represented in text using the normal typing syntax, so pyright represents them the best way it can. You can specify positional-only simple parameters using a list-like syntax. For example, `Y[int, [int, str, float]]` means that the ParamSpec is specialized with three positional-only simple parameters of types `int`, `str`, and `float`. However, PEP 612 doesn't provide a way to specify keyword parameters, `*args`, `**kwargs`, or other more advanced signatures.

[PEP 677](https://peps.python.org/pep-0677/) was an attempt to provide a new syntax for representing callables. Unfortunately, it was rejected by the Python steering council.
